[AIRFLOW-3479] Keeps records in Log Table when delete DAG & refine related tests

XD-DENG opened the pull request:
### Jira

  - https://issues.apache.org/jira/browse/AIRFLOW-3479

### Description

#### Problem This PR Tries to Address

Currently when we delete a DAG (using API or from the UI), it will delete all related records in all tables (all tables in which "dag_id" is available), including "log" table.

However, the records in "log" table should be kept (by default). This would be ideal for multiple reasons, like auditing.

#### What This PR Does

- **When we delete DAG using API**: provide one boolean parameter to let users decide if they want to keep records in Log table when they delete a DAG. Default value it True (to keep records in Log table).

- **When we delete DAG from UI**: will keep records in the Log table when delete records for a specific DAG ID (pop-up message is updated accordingly).


### Tests

Test cases are added to cover:
1. Delete a DAG which doesn't exist.
1. Delete a DAG which is still in DagBag.
1. Successfully delete a DAG + keep records in Log table.
1. Successfully delete a DAG + delete records in Log table as well.

feng-tao commented on tests/api/common/experimental/test_delete_dag.py:
```diff
@@ -0,0 +1,141 @@
+LOG = models.Log
+
+
+class TestDeleteDAG_catch_error(unittest.TestCase):
```
please use capwords for class name(https://www.python.org/dev/peps/pep-0008/#class-names).

XD-DENG replied:
Thanks @feng-tao ! Have addressed this.

feng-tao commented on tests/api/common/experimental/test_delete_dag.py:
```diff
@@ -0,0 +1,141 @@
+            )
+
+
+class TestDeleteDAG_successful_delete(unittest.TestCase):
```
same

XD-DENG replied:
Addressed. Thanks!

feng-tao commented on airflow/api/common/experimental/delete_dag.py:
```diff
@@ -41,6 +49,8 @@ def delete_dag(dag_id):
     # noinspection PyUnresolvedReferences,PyProtectedMember
     for m in models.Base._decl_class_registry.values():
         if hasattr(m, "dag_id"):
+            if keep_records_in_log and m.__name__ == 'Log':
```
does it work with the log is in remote(like s3)? And will it have two different behaviour(api vs webserver) when deleting the log?

XD-DENG replied:
Hi @feng-tao , regarding your questions:

1. `delete_dag()` function does not touch the log files (plain text files in `/log` folder) at all. It only deletes records of given dag_id in the tables in the backend database, including "dag", "task_instance", "dag_run", etc. (actually all tables in which `dag_id` is available). Earlier it also deletes the records in "log" table in the database, which may not be ideal (logs should not be touched by anyone other than Admin). This is what this PR is trying to solve.

1. The behaviour (API vs webserver/UI) will be consistent: when a user deletes a DAG from UI, what happens under the hood is that the URL of the button will invoke the API endpoint with the _default_ parameter (`keep_records_in_log=True`). The only difference is that users can delete records in `log` table using API if they want (of course they need to explicitly specify `keep_records_in_log=False`).

Please let me know if you need any other clarification. Thanks for reviewing!

feng-tao replied:
thanks for the explanation. make sense

XD-DENG replied:
Thanks @feng-tao . May you help re-trigger the Travis test?

Earlier most my test cases failed due to un-related tests. I believe it's some sort of transient errors.

Thanks!

XD-DENG committed:
[AIRFLOW-3479] Keeps records in Log Table when delete DAG+refine tests

 Users will use either API or web UI to delete DAG (after DAG file is
removed):

- Using API: provide one boolean parameter to let users
             decide if they want to keep records in Log table
             when they delete a DAG.
             Default value it True (to keep records in Log table).
- From UI: will keep records in the Log table when delete records for a
           specific DAG ID (pop-up message is updated accordingly).

Tests added to cover:
- Delete a DAG which doesn't exist.
- Delete a DAG which is still in DagBag.
- Successfully delete a DAG + keep records in Log table.
- Successfully delete a DAG + delete records in Log table as well.
